### PR TITLE
Jkantor/submission bff

### PR DESCRIPTION
**Pull request by jansenk**
**TL;DR -** Adds submission / response endpoints for the eventual ORA MFE BFF

JIRA: [AU-1309](https://2u-internal.atlassian.net/browse/AU-1309)

**What changed?**

- Refactored the existing "save draft" and "submit response" logic out of handlers and into api methods
- Refactored error status code to use more custom exceptions
- Wrote BFF handlers and serializers for draft save, submission create, and response load endpoints.
- The ORA submission parsing code can now automatically generate URLs

How to read this PR:
The first commit is a fairly self-contained change to allow ORA submission answers to generate their own URLs.
The second commit is a big one I should have broken up more:

I pulled the "legacy/actions" submission code into a shared "apis/submissions/submissions_actions" folder. It now raises exceptions for exceptional cases rather than returning specific messages and codes.
The "legacy/handlers_mixin" is now a bit less than a shell - I moved request-level logic into these functions, and they call out to the submissions_actions
The MFE view also handles request-level logic and error handling, and then calls down to the submissions_actions.
My thought was basically to have handlers exist to only handle request-level stuff, and all real logic was handed down to submissions_actions
I'm not 100% on pulling "submissions actions" out of "submissions api", but I really wanted, for testability, for submissions_actions to be completely stateless and only be passed data. Mocking turned out to be a challenge for the stateful API object we have. I think some renaming may be in order.
Probably other things I should mention but I'm too hungry to think rn

FYI: @openedx/content-aurora

### CODE REVIEW

File: openassessment/xblock/data_layer/data_layer_mixin.py [hunk 1]
```diff
@@ -3,13 +3,167 @@
...
+                        "download_url": (url)
+                        "description": (str)
+                        "name": (str)
+                        "size": (int)
+                    }
+                ]
+            }
+
+        Raises:
+            submissions.errors.SubmissionError: If there was an error loading the submission
```
Comment (jansenk) [inline openassessment/xblock/data_layer/data_layer_mixin.py:53]: this comment is now incorrect

File: openassessment/xblock/ui_mixins/mfe/mixin.py [hunk 2]
```diff
@@ -4,12 +4,175 @@
...
+        )
 
 
 class MfeMixin:
     @XBlock.json_handler
     def get_block_info(self, data, suffix=""):  # pylint: disable=unused-argument
         block_info = OraBlockInfoSerializer(self)
         return block_info.data
+
+    def get_submission(self, submission_uuid):
```
Comment (jansenk) [inline openassessment/xblock/ui_mixins/mfe/mixin.py:48]: this probably belongs in the submissions_api

File: openassessment/xblock/ui_mixins/mfe/test_mfe_mixin.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+def assert_error_response(response, status_code, error_code, context=''):
+    assert response.status_code == status_code
+    assert response.json['error'] == {
+        'error_code': error_code,
+        'error_context': context
+    }
+
+
+def assert_called_once_with_helper(mock, expected_first_arg, expected_additional_args):
```
Comment (nsprenkle) [inline openassessment/xblock/ui_mixins/mfe/test_mfe_mixin.py:29]: Nit: could be named better, minimally. Bonus points for the cleanliness of having one "expected_args" item instead of split "first" and "additional" args.

```suggestion
def assert_called_once_with_args(mock, expected_args):
```

**Review by nsprenkle — COMMENTED** [review]
Overall, looks good! The one blocking piece is that the serializer fields should be `camelCase` instead of `snake_case`. This is my bad, since I didn't have that in the contracts.

Have proposed a few other small cleanups but nothing else blocking.

File: openassessment/xblock/ui_mixins/mfe/test_mfe_mixin.py (hunk 3 above)
Comment (jansenk) [reply openassessment/xblock/ui_mixins/mfe/test_mfe_mixin.py:29]: I should add a docstring and change the name of additional_args. I don't really want to do a check of all args, I just want to check that the first arg is a specific value and there are the expected number of additional args. I don't see any value in checking the equality of data api wrapper objects, and I think we're constructing different wrapper objects anyway, making a shallow equality test meaningless

File: openassessment/xblock/files_mixin.py [hunk 4]
```diff
@@ -92,12 +92,11 @@ def get_download_urls_from_submission(cls, submission):
...
         raw_answer = submission.get('answer')
         answer = OraSubmissionAnswerFactory.parse_submission_raw_answer(raw_answer)
-        for file_upload in answer.get_file_uploads(missing_blank=True):
-            file_download_url = cls._get_url_by_file_key(file_upload.key)
-            if file_download_url:
+        for file_upload in answer.get_file_uploads(missing_blank=True, generate_urls=True):
+            if file_upload.url:
                 urls.append(
                     file_upload_api.FileDescriptor(
-                        download_url=file_download_url,
```
[the remark this answers is not in the document]
Comment (jansenk) [reply openassessment/xblock/files_mixin.py:100]: No, it's the _download url_ of an _uploaded_ file. Confusing, but I'm sure this is right

File: openassessment/xblock/leaderboard_mixin.py [hunk 5]
```diff
@@ -77,7 +77,7 @@ def render_leaderboard_complete(self, student_item_dict):
 
         # Retrieve top scores from the submissions API
         # Since this uses the read-replica and caches the results,
-        # there will be some delay in the request latency.
+        # there will be some delay in the  latency.
```
[the remark this answers is not in the document]
Comment (jansenk) [reply openassessment/xblock/leaderboard_mixin.py:80]: how the heck did you catch that 😮

File: openassessment/xblock/leaderboard_mixin.py [hunk 6]
```diff
@@ -88,12 +88,11 @@ def render_leaderboard_complete(self, student_item_dict):
...
             score['files'] = []
-            for uploaded_file in answer.get_file_uploads(missing_blank=True):
-                file_download_url = self._get_file_download_url(uploaded_file.key)
-                if file_download_url:
+            for uploaded_file in answer.get_file_uploads(missing_blank=True, generate_urls=True):
+                if uploaded_file.url:
                     score['files'].append(
                         file_upload_api.FileDescriptor(
-                            download_url=file_download_url,
+                            download_url=uploaded_file.url,
```
[the remark this answers is not in the document]
Comment (jansenk) [reply openassessment/xblock/leaderboard_mixin.py:95]: Yes, I believe this is right

File: openassessment/xblock/apis/submissions/submissions_actions.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
+from openassessment.workflow.errors import AssessmentWorkflowError
+from openassessment.xblock.utils.data_conversion import (
+    format_files_for_submission,
+    prepare_submission_for_serialization,
+)
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+def submit(data, block_config_data, block_submission_data, block_workflow_data):
```
[the remark this answers is not in the document]
Comment (jansenk) [reply openassessment/xblock/apis/submissions/submissions_actions.py:29]: I agree, but imo this slight messiness is okay for now. If we want to do later refactors in this direction I'm all for it, but I figure we should get everything to this-ish level, then move forward from there if we want to

File: openassessment/xblock/ui_mixins/mfe/mixin.py [hunk 8]
```diff
@@ -4,12 +4,175 @@
...
+            raise OraApiException(500, ErrorCodes.UNKNOWN_ERROR, str(e)) from e
+
+    @XBlock.json_handler
+    def submission(self, data, suffix=""):
+        if suffix == 'draft':
+            return self._submission_draft(data)
+        elif suffix == "create":
+            return self._submission_create(data)
+        else:
+            raise OraApiException(404, ErrorCodes.UNKNOWN_SUFFIX)
```
[the remark this answers is not in the document]
Comment (jansenk) [reply openassessment/xblock/ui_mixins/mfe/mixin.py:116]: From the api contract doc

File: openassessment/xblock/ui_mixins/mfe/mixin.py [hunk 9]
```diff
@@ -4,13 +4,32 @@
...
+    SubmissionValidationException,
+    SubmitInternalError
 )
+from openassessment.xblock.ui_mixins.mfe.constants import ErrorCodes
+from openassessment.xblock.ui_mixins.mfe.ora_config_serializer import OraBlockInfoSerializer
+from openassessment.xblock.ui_mixins.mfe.page_context_serializer import PageDataSerializer
+
+
+class OraApiException(JsonHandlerError):
+    def __init__(self, status_code, error_code, error_context=''):
```
Comment (nsprenkle) [inline openassessment/xblock/ui_mixins/mfe/mixin.py:25]: Nit: can you throw a docstring on here?

File: openassessment/xblock/ui_mixins/mfe/ora_config_serializer.py [hunk 10]
```diff
@@ -221,3 +222,20 @@ def get_baseAssetUrl(self, block):
 
     def get_prompts(self, block):
         return [prompt["description"] for prompt in block.prompts]
+
+
+class GetBlockLearnerSubmissionDataSerializer(Serializer):
```
Comment (nsprenkle) [inline openassessment/xblock/ui_mixins/mfe/ora_config_serializer.py:227]: Nit: naming? Does it need that many qualifiers?

```suggestion
class ResponseDataSerializer(Serializer):
```

File: openassessment/xblock/ui_mixins/mfe/mixin.py [hunk 11]
```diff
@@ -30,3 +49,83 @@ def get_block_learner_assessment_data(self, data, suffix=""):  # pylint: disable
...
+            submissions_actions.save_submission_draft(student_submission_data, self.config_data, self.submission_data)
+        except KeyError as e:
+            raise OraApiException(400, ErrorCodes.INCORRECT_PARAMETERS) from e
+        except SubmissionValidationException as exc:
+            raise OraApiException(400, ErrorCodes.INVALID_RESPONSE_SHAPE, str(exc)) from exc
+        except DraftSaveException as e:
+            raise OraApiException(500, ErrorCodes.INTERNAL_EXCEPTION) from e
+
+    def _submission_create(self, data):
+        from submissions import api as submission_api
```
Comment (nsprenkle) [inline openassessment/xblock/ui_mixins/mfe/mixin.py:65]: Nit: more descriptive names would be helpful here for readability

```suggestion
    def _save_draft_response_and_wrap_error(self, data):
        try:
            student_submission_data = data['response']['text_responses']
            submissions_actions.save_submission_draft(student_submission_data, self.config_data, self.submission_data)
        except KeyError as e:
            raise OraApiException(400, ErrorCodes.INCORRECT_PARAMETERS) from e
        except SubmissionValidationException as exc:
            raise OraApiException(400, ErrorCodes.INVALID_RESPONSE_SHAPE, str(exc)) from exc
        except DraftSaveException as e:
            raise OraApiException(500, ErrorCodes.INTERNAL_EXCEPTION) from e

    def _submit_response_and_wrap_error(self, data):
        from submissions import api as submission_api
```

**Review by nsprenkle — APPROVED** [review]
Looks good, with some nits for naming & documentation clarity

File: openassessment/staffgrader/tests/test_get_submission_info.py [hunk 12]
```diff
@@ -39,9 +39,11 @@ def _mock_parse_submission_raw_answer(self, **kwargs):
         ) as mock_parse:
             yield mock_parse
 
-    def _mock_get_download_urls_from_submission(self, xblock, **kwargs):
+    @contextmanager
+    def _mock_get_download_urls(self, **kwargs):
         """ Helper method to mock the get_download_urls_from_submission method """
```
Comment (leangseu-edx) [inline openassessment/staffgrader/tests/test_get_submission_info.py:44]:
```suggestion
        """ Helper method to mock the get_download_urls method """
```

File: openassessment/xblock/files_mixin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (leangseu-edx) [reply openassessment/xblock/files_mixin.py:100]: There are currently 4 types of backend `s3, filesystem, swift, django`. It seems `django-storage` has different implementation for each of those. And s3 would generate different url. It has different timeout value.

File: openassessment/xblock/leaderboard_mixin.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (leangseu-edx) [reply openassessment/xblock/leaderboard_mixin.py:95]: ditto from above.

**Review by leangseu-edx — CHANGES_REQUESTED** [review]
Look good beside the download url/upload url.

File: openassessment/xblock/ui_mixins/mfe/mixin.py (hunk 11 above)
Comment (jansenk) [reply openassessment/xblock/ui_mixins/mfe/mixin.py:73]: I don't think that error wrapping is really something that needs to be specified in these method names, but I will try to make these names more descriptive

**Review by leangseu-edx — APPROVED** [review]
LGTM
